### PR TITLE
#881 Fixed behavior in case of wrong file upload

### PR DESCRIPTION
**Pull request by exper0**
#881 Fixed behavior in case of wrong file upload

### CODE REVIEW

**Comment by karato** [comment]
@exper0 Thanks, someone will review your pull request soon

**Comment by karato** [comment]
@dmzaytsev the pull request is rather small, pls review it

**Comment by dmzaytsev** [comment]
@exper0 what's about a test?

**Comment by exper0** [comment]
@dmzaytsev the issue is specific to DyBase - so should I create integration test?

**Comment by dmzaytsev** [comment]
@exper0 I think you can create a new test in `TkAttachTest`
try to attach a zero-length file and check for the empty bout attachments list

**Comment by exper0** [comment]
@dmzaytsev it's not possible to create unit test with DyBase. Correct me if I'm wrong

**Comment by exper0** [comment]
@dmzaytsev so in order to test it I have to create new integration test (probably in DyAttachmentsITCase)  or even create a new pack

**Comment by dmzaytsev** [comment]
@exper0 look at `TkAttachTest#sendsMessageToBout()`

**Comment by dmzaytsev** [comment]
@exper0 let's create a test for TkAttach first
you can create a puzzle to create a test in `DyAttachmentsITCase` as well

**Comment by dmzaytsev** [comment]
@exper0 you made some changes in TkAtatch so we need to check it by a unit test

**Comment by exper0** [comment]
@dmzaytsev This will not work for this case.
1) #881 is about DyBase implementation and about DyAttachment only. Doing similar to `TkAttachTest#sendsMessageToBout()` will not test anything - it will pass even without my fix. Only DyAttachment implementation contains limitation on empty file (throws exception). 
2) DyBase can be instantiated on integration test - becaise Dynamo DB is complex database that requires ports reservation etc. All such cases gathered in a set of *ITCase classes and runs on integration test phase during maven build

**Comment by exper0** [comment]
@dmzaytsev look [at this](https://github.com/yegor256/netbout/blob/master/netbout-web/src/main/java/com/netbout/dynamo/DyAttachment.java#L213)

**Comment by dmzaytsev** [comment]
@exper0 ok, let's create an integration test

**Comment by exper0** [comment]
@dmzaytsev refactored the code and moved validations to TkAttach. It's more consistent to have such logic common across all attachment implementations. Added unit test

File: netbout-web/src/main/java/com/netbout/dynamo/DyAttachment.java [hunk 1]
```diff
@@ -188,40 +187,7 @@ public InputStream read() throws IOException {
     @Override
     public void write(final InputStream stream, final String ctype,
         final String etag) throws IOException {
-        if (etag.isEmpty()) {
```
Comment (dmzaytsev) [inline netbout-web/src/main/java/com/netbout/dynamo/DyAttachment.java]: @exper0 why did you remove `etag` and `ctag` validation ?
I hadn't saw these validations in TkAttach

File: netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java [hunk 2]
```diff
@@ -96,38 +96,44 @@ public Response act(final Request req) throws IOException {
...
+            final String ctype = TkAttach.ctype(temp);
+            msg.append(" (").append(temp.length())
+                .append(" bytes, ").append(ctype).append(')');
+            try (InputStream is = new FileInputStream(temp)) {
+                bout.attachments().get(name).write(
+                    is,
+                    ctype, Long.toString(System.currentTimeMillis())
+                );
+            }
+            bout.messages().post(msg.toString());
```
Comment (dmzaytsev) [inline netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java]: @exper0 please explain why did you add this line?

Comment (dmzaytsev) [inline netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java]: @exper0 sorry, skip my previous comment

File: netbout-web/src/test/java/com/netbout/rest/bout/TkAttachTest.java [hunk 3]
```diff
@@ -137,6 +153,57 @@ public void ignoresWrongRequest() throws Exception {
...
+            MatcherAssert.assertThat(
+                "file unexpectedly exists after attach failure",
+                new Attachments.Search(
+                    alias.inbox().bout(bout).attachments()
+                ).exists(file),
+                Matchers.is(false)
+            );
+            return;
+        }
+        Assert.fail("Expected RsFailure exception but nothing was thrown");
```
Comment (dmzaytsev) [inline netbout-web/src/test/java/com/netbout/rest/bout/TkAttachTest.java]: @exper0 let's move this line before `catch`
and remove `return` in catch clause

File: netbout-web/src/test/java/com/netbout/rest/bout/TkAttachTest.java [hunk 4]
```diff
@@ -80,13 +86,23 @@ public void sendsMessageToBout() throws Exception {
             new RqMultipart.Fake(
                 TkAttachTest.fake(number),
                 new RqWithHeaders(
-                    new RqLive(
-                        new ByteArrayInputStream("content1".getBytes())
-                    ),
+                    new Request() {
```
Comment (dmzaytsev) [inline netbout-web/src/test/java/com/netbout/rest/bout/TkAttachTest.java]: @exper0 we can move creation of the request into static method and use it here and in the `ignoresWrongFile`

**Comment by dmzaytsev** [comment]
@exper0 please see my comments

File: netbout-web/src/main/java/com/netbout/dynamo/DyAttachment.java (hunk 1 above)
Comment (exper0) [inline netbout-web/src/main/java/com/netbout/dynamo/DyAttachment.java]: @dmzaytsev TkAttach is the place where ctag and etag are created. It's strange to create them and check straight away. File content and name is something external - so it's the only thing we need to validate

**Comment by exper0** [comment]
@dmzaytsev made anothing commit taking into account your comments

File: netbout-web/src/main/java/com/netbout/dynamo/DyAttachment.java (hunk 1 above)
Comment (dmzaytsev) [inline netbout-web/src/main/java/com/netbout/dynamo/DyAttachment.java]: @exper0 you're right

**Comment by dmzaytsev** [comment]
@exper0 looks good to me, thanks

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by karato** [comment]
@dmzaytsev Thanks for your contribution, **30 mins** was added to your account, payment ID is `75188970`, 21 hours and 43 mins spent total; you're getting extra minutes here (c=15); +30 added to your rating, current score is: [+2235](http://www.netbout.com/b/35166?open=rating)

**Comment by karato** [comment]
@rultor deploy now pls
